Add --files-from and --files-from0 options

rotty opened the pull request:
This makes attic usable with custom file listers (e.g. 'find'
invocations) and pre-assembled file lists. Both
newline-delimited (--files-from) and NUL-delimited (--files-from0)
formats are supported.

rotty committed:
Add --files-from and --files-from0 options

This makes attic usable with custom file listers (e.g. 'find'
invocations) and pre-assembled file lists. Both
newline-delimited (--files-from) and NUL-delimited (--files-from0)
formats are supported.

ThomasWaldmann commented:
could you add some unit tests?

rotty commented:
I'll do so, hopefully in this weekend.

rotty committed:
Add basic unit tests for --file-from and --file-from0

Test the basic functionality of "attic create" with --files-from and
--files-from0 options. Note that the tests are not very exhaustive;
missing in particular is:

- reading from stdin, as the test harness doesn't emulate
  sys.stdin.buffer.
- use of multiple --files-from and/or --files-from0 options.

rotty commented:
I've now added basic unit tests, however, as noted in the commit message, these are not yet very comprehensive. If you decide to merge this branch at this point nevertheless, feel free to squash the two commits (code and unit test).

ThomasWaldmann commented on attic/helpers.py:
```diff
@@ -528,6 +528,63 @@ def make_path_safe(path):
+    def __call__(self, string):
+        result = super().__call__(string)
+        # Work around http://bugs.python.org/issue14156
+        if self._binary and result is sys.stdin or result is sys.stdout:
```
is the line above doing what you want?

rotty replied:
Manual tests using stdin work (this is my primary use case, after all), however, to add automatic tests, one would need to extend 'ArchiverTestCaseBase.attic()' to emulate sys.stdin.buffer to enable binary input from stdin.

ThomasWaldmann replied:
do you mean it like this:  if (self._binary and result is sys.stdin) or result is sys.stdout: ?

rotty replied:
TW notifications@github.com writes:

> > +class FileType(argparse.FileType):
> > +
> > -    [...]
> >   +
> > -    def **call**(self, string):
> > -        result = super().**call**(string)
> > -        # Work around http://bugs.python.org/issue14156
> > -        if self._binary and result is sys.stdin or result is sys.stdout:
> 
> do you mean it like this:  if (self._binary and result is sys.stdin)
> or result is sys.stdout: ?
> 
> Oh, good catch, there is a set of parens missing (should be 'if
> self._binary and (result is sys.stdin or result is
> sys.stdout)'. Obviously, I didn't test with sys.stdout. I'll fix this,
> along with adding a unit test for this class.

Thanks, Rotty

rotty replied:
Another thing that occurred to me, and which i'm not exactly sure what to do about, is that the code in `FileType.__call__()` monkey-patches (i.e., adds attributes to) `sys.stdin` or `sys.stderr`, if they are the result of `super().__call__()`.

I think this monkey-patching is fine for newly-created file objects, but less so for `sys.stdin` and `sys.stdout`. It might however be acceptable in the context of an application (as opposed to a general-purpose library). Do you have an opinion on that topic?

ThomasWaldmann commented on attic/helpers.py:
```diff
@@ -528,6 +528,63 @@ def make_path_safe(path):
+        if start < len(data):
+            bufs.append(data[start:])
+    if len(bufs) > 0:
+        yield empty.join(bufs)
```
please add some unit tests just for this function.
consider edge cases like:
empty input file
file starts/ends with delim
2 delims directly following each other

rotty committed:
Add unit tests for 'iter_delim' helper

rotty committed:
helpers: fix logic in 'FileType', add test

Before this change, the FileType wrapper would always return a binary
file object when used with stdout, regardless of the file mode.

rotty commented:
Oh, that Travis CI failure is sneaky: I can reproduce it locally, but it only happens when running `python3 -m attic.testsuite.run -b`, but not without `-b` ("Buffer stdout and stderr during tests").

rotty committed:
testsuite: work around lack of sys.stdout buffer

Whem the test suite is run with the '-b' option, the tests methods no
longer have sys.stdout.buffer available, as sys.stdout is replaced by a
StringIO instance.
